### PR TITLE
Make CLIP model could use new added tokens with meaningful pooling

### LINKED ISSUE #24650: CLIP pooling is not compatible with adding new tokens
### System Info

Feature request (Duplicate of #21029)

For textual inversion in diffusers, we are adding tokens that have a higher token id than the eos token. So when we get clip embeddings for textual inv tokens, we need to change the pooling so it gets the eos token and not the arg max token.

Motivation

This is an issue that should be fixed as the clip embeddings won't work once we add more tokens to the tokenizer. This hasn't been a huge issue so far because most models use the hidden layers directly but [the new paper on SDXL](https://github.com/Stability-AI/generative-models/blob/main/assets/sdxl_report.pdf) also mentions using the pooled output now. 

@ArthurZucker @younesbelkada 

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Add new token to tokenizer
2. Encode tokens with CLIPTextModel
3. Get pooled output

### Expected behavior

Pooled output considers the added token ids vs eos id instead of argmax

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Fix #24650

This is to address the feature request #24650.

Although we have default values of bos/eos been corrected in #24773, the existing config on the Hub still have the incorrect value `1` and `2`, which prevents CLIP model to use new added tokens when a user add them.

Although we can open mass PRs on the Hub, I want to decouple (slightly) this with the ability to support such feature.
With this PR, if a user want to use new added tokens, they has to specify/update the `eos_token_id`.

**We don't need to wait all Hub repo. to be updated to merge this PR.**

### CODE REVIEW

File: src/transformers/models/clip/modeling_clip.py
Comment (ydshieh) [inline src/transformers/models/clip/modeling_clip.py:756]: The if/else condition should be enough.

I don't think there is any user changing this from `2` to anything that is not `vocab_size - 1` (the `vocab_size` of the original clip tokenizer)

File: src/transformers/models/clip/modeling_clip.py
Comment (ydshieh) [inline src/transformers/models/clip/modeling_clip.py:771]: As the (default) pad and eos have the same token id in CLIP  config. We need to get the `1st` position of that value in `input_ids`. 

`input_ids == self.eos_token_id` gives where we have them, then argmax (on 0/1 values) to get the 1st position.

Comment (ydshieh) [reply src/transformers/models/clip/modeling_clip.py:771]: I need to update the TF/Flax code too (as the model tester has no eos_token_ids specified and go through this new block in torch, but not yet in the TF/Flax version)

File: src/transformers/models/clip/modeling_clip.py
Comment (patil-suraj) [reply src/transformers/models/clip/modeling_clip.py:756]: This looks good, but I'm always a bit afraid of hardcoding values like this, as we never know how the users may have modified the config. I will defer to @amyeroberts for this :)

File: src/transformers/models/clip/modeling_clip.py
Comment (patil-suraj) [inline src/transformers/models/clip/modeling_clip.py:775]: very cool!

Comment (ydshieh) [reply src/transformers/models/clip/modeling_clip.py:775]: >  we never know how the users may have modified the config.

A possibility is to check if the `eos_token_id` is greater than or equal to the default value in the config (40257)

  - no: it may not be 2 (as user may changed it), but we are sure nothing we can't do and should keep the original logic
  - yes: we can use the new block added in this PR

File: src/transformers/models/clip/modeling_clip.py
Comment (amyeroberts) [reply src/transformers/models/clip/modeling_clip.py:756]: We're all a bit afraid BUT I think it's OK 😬 

I think you made a convincing point before, @patil-suraj. This token isn't used by the model anywhere (yet), and `eos_token_id=2` is wrong. There's three possible cases: 

* `eos_token_id=2` - then the old behaviour is still preserved
* `eos_token_id=vocab_size - 1` - then the model behaviour is still preserved (and now correct)
* `eos_token_id=something_else`. As it's not used by the model, it doesn't really make sense to update this unless the tokenizer is also updated with a new `eos_token_id` and then finetuned. There's a small chance this has happened, however I think we would have had this `eos_token_id` inconsistency flagged beforehand in this case and the previous pooling behaviour would have been incorrect. 

I propose waiting until after the branch is cut, and then merging. WDTY?

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! 

I think it's OK for us to add this and provides a solid way to remain backwards compatible with the old config. 

I left a comment, but given the possible change in model behaviour, let's wait to merge until after the next release on Monday

File: src/transformers/models/clip/modeling_clip.py
Comment (ydshieh) [reply src/transformers/models/clip/modeling_clip.py:756]: I can wait the branch cut 😄 !

> eos_token_id=something_else

The chance is very small, so I won't spend time to change it to `if eos_token_id < default config value (the now `49407`)`.

(as this needs to inspect the __init__ argument of a class)

**Comment by ydshieh** [comment]
I will merge once the branch is cut tonight.

**Comment by ydshieh** [comment]
Sorry for the spam: @sgugger said the branch cut would be on next Monday. I think it's safer to wait until then.

**Pull request merged**
